### PR TITLE
feat(cli): Improve management of --map

### PR DESCRIPTION
**Pull request by MathiusD**
If `--map` isn't specified, show all id of maps available

```
$ ./target/release/vakfu --path ~/.config/Ankama/zaap/wakfu
Error: Map isn't specified, following maps id are available :
[9, 99, 147, 149, 150, 151, 153, 154, 163, 165, 179, 184, 189, 202, 209, 219, 223, 224, 225, 229, 231, 274, 295, 296, 297, 299, 301, 309, 311, 312, 319, 323, 325, 327, 329, 330, 337, 338, 339, 341, 342, 343, 346, 347, 348, 349, 350, 351, 358, 359, 360, 363, 365, 368, 369, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 436, 437, 438, 439, 440, 441, 443, 444, 445, 446, 447, 448, 449, 452, 453, 454, 455, 456, 458, 459, 460, 461, 466, 467, 468, 471, 472, 474, 475, 476, 477, 479, 481, 483, 485, 489, 493, 495, 496, 499, 504, 507, 508, 509, 510, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526, 527, 528, 529, 532, 533, 534, 535, 536, 537, 539, 540, 541, 542, 543, 544, 545, 546, 548, 549, 552, 553, 554, 557, 559, 560, 562, 569, 571, 572, 573, 579, 581, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 597, 598, 599, 600, 601, 602, 603, 604, 609, 610, 611, 615, 616, 617, 618, 619, 620, 622, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 651, 652, 653, 654, 655, 656, 657, 658, 659, 660, 661, 662, 663, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676, 677, 678, 679, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 699, 704, 705, 709, 711, 712, 713, 714, 715, 716, 717, 718, 719, 720, 723, 725, 726, 728, 729, 730, 731, 732, 733, 734, 735, 736, 737, 738, 739, 740, 741, 742, 743, 744, 747, 748, 749, 750, 751, 752, 753, 754, 755, 756, 757, 758, 759, 760, 761, 762, 763, 764, 765, 766, 767, 768, 769, 770, 771, 772, 773, 774, 775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 792, 793, 794, 795, 796, 798, 799, 800, 801, 802, 803, 804, 805, 806, 807, 808, 809, 813, 814, 815, 816, 817, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 831, 832, 833, 837, 838, 839, 840, 841, 842, 843, 844, 845, 847, 848, 849, 850, 851, 853, 854, 855, 856, 858, 859, 860, 862, 863, 864, 865, 866, 867, 868, 869, 871, 872, 873, 874, 875, 876, 879, 885, 886, 887, 888, 889, 891, 892, 895, 896, 898, 899, 901, 902, 907, 909, 910, 911, 912, 913, 916, 917, 918, 919, 920, 922, 923, 924, 925, 926, 929, 930, 931, 932, 933, 934, 935, 936, 937, 938, 939, 940, 941, 943, 944, 945, 946, 947, 948, 949, 951, 952, 953, 954, 955, 957, 959, 960, 961, 962, 963, 965, 967, 968, 969, 970, 971, 972, 973, 974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999, 1000, 1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1011, 1012, 1013, 1016, 1017, 1018, 1019, 1020, 1024, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1037, 1038, 1039, 1040, 1042, 1043, 1044, 1045, 1046, 1047, 1048, 1050, 1051, 1052, 1053, 1054, 1056, 1057, 1061, 1062, 1063, 1069, 1071, 1072, 1073, 1074, 1075, 1076, 1079, 1080, 1081, 1082, 1083, 1084, 1086, 1087, 1089, 1090, 1091, 1092, 1093, 1094, 1095, 1096, 1097, 1098, 1099, 1100, 1101, 1102, 1103, 1104, 1105, 1106, 1107, 1108, 1109, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1120, 1121, 1122, 1123, 1124, 1126, 1127, 1128, 1129, 1130, 1131, 1132, 1133, 1134, 1135, 1136, 1137, 1138, 1139, 1140, 1141, 1143, 1144, 1145, 1146, 1147, 1149, 1150, 1151, 1153, 1154, 1155, 1156, 1157, 1158, 1159, 1160, 1161, 1163, 1165, 1169, 1179, 1184, 1201, 1202, 1203, 1204, 1206, 1207, 1208, 1209, 1210, 1211, 1212, 1213, 1214, 1215, 1216, 1217, 1218, 1219, 1220, 1221, 1222, 1223, 1224, 1225, 1226, 1227, 1228, 1229, 1231, 1232, 1233, 1234, 1235, 1236, 1237, 1238, 1239, 1240, 1241, 1242, 1243, 1244, 1245, 1246, 1247, 1248, 1249, 1250, 1251, 1252, 1253, 1254, 1255, 1256, 1257, 1258, 1259, 1260, 1261, 1262, 1263, 1264, 1265, 1266, 1267, 1268, 1269, 1270, 1271, 1272, 1273, 1274, 1275, 1276, 1277, 1278, 1279, 1280, 1281, 1282, 1283, 1284, 1285, 1286, 1287, 1288, 1289, 1290, 1291, 1292, 1293, 1294, 1295, 1296, 1297, 1298, 1299, 1300, 1301, 1302, 1303, 1304, 1305, 1306, 1307, 1308, 1309, 1310, 1311, 1312, 1313, 1314, 1315, 1316, 1317, 1318, 1319, 1320, 1321, 1322, 1323, 1324, 1325, 1326, 1327, 1328, 1329, 1330, 1331, 1332, 1333, 1334, 1335, 1336, 1337, 1338, 1339, 1340, 1341, 1342, 1343, 1344, 1345, 1346, 1347, 1348, 1349, 1350, 1351, 1352, 1353, 1354, 1355, 1356, 1357]
```

### CODE REVIEW

File: src/utils/id.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                    }
+                    None => {}
+                };
+            }
+            Err(_e) => continue,
+        };
+    }
+    maps_id.sort();
+    return Ok(maps_id);
+}
```
Comment (jac3km4) [inline src/utils/id.rs]: these pattern matches could all be removed by just using the `?` operator, or you could rewrite it with an iterator like this:
```suggestion
use std::fs;
use std::path::Path;

use anyhow::Result;
use itertools::Itertools;

pub fn get_maps_id(map_dir: impl AsRef<Path>) -> Result<Vec<i32>> {
    let mut maps_id: Vec<_> = fs::read_dir(map_dir)?
        .map_ok(|entry| entry.path())
        .filter_ok(|entry| entry.extension() == Some("jar".as_ref()))
        .filter_map_ok(|entry| Some(entry.file_stem()?.to_str()?.parse()))
        .flatten_ok()
        .try_collect()?;
    maps_id.sort();
    Ok(maps_id)
}
```

Comment (jac3km4) [reply src/utils/id.rs]: also I'd rename this to `get_map_ids`

File: src/main.rs [hunk 2]
```diff
@@ -1,65 +1,85 @@
...
+                        "Map isn't specified, following maps id are available :\n{:?}",
+                        maps_id
+                    )),
+                    Err(_e) => Err(anyhow!(
+                        "Map isn't specified, but no maps found in game_path specified."
+                    )),
+                }
+            }
+            _ => Err(anyhow!(_e)),
+        },
```
Comment (jac3km4) [inline src/main.rs]: there's no need for a nested `match`, you can unnest it:
```suggestion
        Err(Error::MissingOption(_) | Error::OptionWithoutAValue("map")) => {
            match get_map_ids(maps_path.join("gfx")) {
                Ok(map_ids) => Err(anyhow!(
                    "Map isn't specified, following maps id are available: \n{:?}",
                    map_ids
                )),
                Err(_e) => Err(anyhow!(
                    "Map isn't specified, but no maps found under provided game_path"
                )),
            }
        }
        Err(other) => Err(anyhow!(other)),
```

**Comment by jac3km4** [comment]
nice addition, I left 2 suggestions

**Comment by MathiusD** [comment]
Thanks a lot for the suggests, I just corrected it!

File: src/main.rs (hunk 2 above)
Comment (MathiusD) [reply src/main.rs]: Thanks, I didn't see how to write it at first (I'm a beginner in Rust and I think it's noticeable)

File: src/utils/id.rs (hunk 1 above)
Comment (MathiusD) [reply src/utils/id.rs]: Renamed and corrected the name of the function (I hadn't had the opportunity to manipulate itertools before and this allows me to do so, thanks for showing it to me!)

**Comment by jac3km4** [comment]
lgtm!

**Pull request merged**